include/drivers/gpio.h

```diff
@@ -30,14 +30,57 @@ extern "C" {
  * @brief GPIO Driver APIs
  * @defgroup gpio_interface GPIO Driver APIs
  * @ingroup io_interfaces
+ *
+ * @{
+ * @name GPIO Atomicity and Interrupt-Related Behavior
+ *
  * @{
+ * All GPIO peripheral API operations satisfy the following promises:
+ *
+ * * All functions may be invoked from interrupt service routines and
+ *   callbacks.  Functions that cannot fulfill their behavioral
+ *   requirements in that context may return `-EWOULDBLOCK`.
+ * * All pin configuration functions must be atomic.
```

Comment:
Do you think `atomic` reflects the fact that we (will) disable interrupts as opposed to using another synchronization mechanism? I think it does, and it is sort of implied by the fact that all functions can be called from an ISR, but just checking.

Reply:
I think `atomic` means what I said it means in #18970.  Which we should reach agreement on and document.

Reply:
Fair enough. Let's use the terms as they are in #18970 and get a PR in for that.